gremlin-go/driver/authInfo.go

```diff
@@ -55,3 +66,30 @@ func BasicAuthInfo(username string, password string) *AuthInfo {
 func HeaderAuthInfo(header http.Header) *AuthInfo {
 	return &AuthInfo{Header: header}
 }
+
+// DynamicAuth is an AuthInfoProvider that allows dynamic credential generation.
+type DynamicAuth struct {
+	fn func() AuthInfoProvider
+}
+
+var (
+	_ AuthInfoProvider = (*DynamicAuth)(nil)
+
+	// NoopAuthInfo is a no-op AuthInfoProvider that can be used to disable authentication.
+	NoopAuthInfo = NewDynamicAuth(func() AuthInfoProvider { return &AuthInfo{} })
```

Comment:
This is implemented as dynamic auth so that the returning value can't be changed, e.g. by setting the public fields of the AuthInfo struct after use.